### DISCUSSION TITLE
Memory leakage

### DISCUSSION THREAD

**Discussion by lixiaoyong12** [apache/airflow#29264] [General]
### Apache Airflow version

2.5.1

### What happened

Linear memory growth，The system（200 DAG） runs for 50 minutes and the memory is increased from 100 MB to 500 MB.

![image](https://user-images.githubusercontent.com/34184204/215723671-774b1ed3-f92a-4d73-a296-66934e98a5e4.png)

Comment out the following code. The memory does not increase.

![image](https://user-images.githubusercontent.com/34184204/215723776-97db7de4-1bd5-480d-900b-21cb291a209c.png)

### How to reproduce

Run 200 2-minute dag files and run the ps command to check the memory usage.

### Operating System

euleros

### Versions of Apache Airflow Providers

2.5.0

### Deployment

Other

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Let me convert it to discussion. There are far too little details in this request to be able to act on it.

This is absolutely normal that (for example cache) memory grows in linux when the system is running and which process leaks memory. This could be - for example - your database allocating more (cache) memory as it is inserting and caching things in the DB. 

You have not explained what kind of memory is growing nor which process leaks . 

This is really not actionable report until more details is given and showing which process leaks "real" memory.  You can see similar issues here: https://github.com/apache/airflow/discussions/28597

**Reply by potiuk** [reply]
Please do more analysis to show that this is a real problem (I believe it is not) and what is leaking memory (and what kind of memory to show that this is really a leak).

**Reply by lixiaoyong12** [reply]
When the same task is executed, the memory of the airflow scheduler in 2.0.1 is 70 MB to 80 MB. However, the memory of the airflow scheduler in 2.5.0 keeps increasing and reaches 500 MB in 50 minutes. The memory allocated to the airflow scheduler is 500 MB. If the size is greater than 500 MB, the kill operation is performed. （ps -aux | grep scheduler）

**Reply by lixiaoyong12** [reply]
The analysis code is caused by the association of multiple tables. When the task is generated, the memory keeps increasing. However, I have not found a good solution.

**Reply by potiuk** [reply]
Did yout try to increase the memory and see if it flattens out? I think simply this query might require more memory. For every - even smallest installation of Airflow we recommend at least 4GB RAM even to begin with. Maybe simply you gave Airflow too little memory. Increase the memory to say 4GB and see if you have the same problem. 

Also this seems to be related to sqlalchemy code, so please double check if you are using the version of sqlalchemy that is recommended in the constraints of Airflow. See https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html on how to install Airflow with the dependencies of the "golden" set of dependencies.

It might very well be that you are using some very old version of sqlalchemy with the memory like. Please make sure you reinstall airflow in the way we recommend (with constraints) and see if you can reproduce it there.

**Reply by potiuk** [reply]
CC: @uranusjr, @ashb  Early warnings if the changes/response from @lixiaoyong12 will indicate that the issue persist and it will continue to show the same behaviours.

It **might** be **just** an effect of bulk_insert_mapping allocating a lot of memory (though it seems a bit excessive with the behaviour reported and the fact that it grows continuously over 50 minutes.

I found this https://groups.google.com/g/sqlalchemy/c/pBwh9coOH78. which is an indication that there might be huge spikes in memory use for bulk_insert_mapping but they should not be left behind (unless we have a problem where Task Instances are held in memory and not garbage-collected).

**Reply by lixiaoyong12** [reply]
1) The sqlalchemy version I use is 1.4.44, and the installation is based on the airflow installation guide.  
2) When the task_instance table is not associated with the other table, the same data is inserted in batches. The size of the data is about 140 MB.
3) The bulk_save_objects, add, and add_all operations are also performed. The memory increases only when data is inserted into the task_instance object.

**Reply by lixiaoyong12** [reply]
I don't task with 4 GB of memory, but my tasks should be scheduled at 150 MB of memory.

**Comment by lixiaoyong12** [comment]
I solved the problem with https://github.com/sqlalchemy/sqlalchemy/blob/main/doc/build/core/connections.rst ， thanks you @potiuk

**Reply by potiuk** [reply]
I wonder what was the solution  @lixiaoyong12 ?

**Reply by lixiaoyong12** [reply]
supports_statement_cache = False

**Reply by potiuk** [reply]
what dialect do you have (I have not seen. you mentioned it)?
